Mathlib/RingTheory/Nilpotent/GeometricallyReduced.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2025 Dion Leijnse. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dion Leijnse
+-/
+import Mathlib.Algebra.Lie.OfAssociative
+import Mathlib.FieldTheory.IsAlgClosed.AlgebraicClosure
+import Mathlib.LinearAlgebra.TensorProduct.Finiteness
+import Mathlib.RingTheory.Flat.FaithfullyFlat.Basic
+import Mathlib.RingTheory.Henselian
+
+/-!
+# Geometrically reduced algebras
+
+In this file we introduce geometrically reduced algebras.
+For a field `k`, we say that a `k`-algebra `A` is geometrically reduced (`IsGeometricallyReduced`)
+if the tensor product `A ⊗[k] AlgebraicClosure k` is reduced.
+
+## Main results
+- `all_FG_geometricallyReduced_isGeometricallyReduced`: if all finitely generated subalgebras of `A`
+  are geometrically reduced, then `A` is geometrically reduced.
+
+## References
+- See [https://stacks.math.columbia.edu/tag/05DS] for some theory of geometrically reduced algebras.
+  Note that their definition differs from the one here, we still need a proof that these are
+  equivalent.
+
+-/
+
+noncomputable section
+variable (k : Type) (A : Type) [Field k] [CommRing A] [Algebra k A]
+variable (B : Type) [CommRing B] [Algebra k B]
+
+
+/-- Given a morphism f : A →ₐ[k] B of k-algebras and a k-algebra K, the basechange of f by the
+identity map K → K -/
+def baseChange (f : A →ₐ[k] B) (K : Type) [CommRing K] [Algebra k K] :
+    TensorProduct k K A →ₐ[K] TensorProduct k K B :=
+    Algebra.TensorProduct.map (AlgHom.id K K) f
+
+
+lemma baseChange_by_field_inj (f : A →ₐ[k] B) (h : Function.Injective f) (K : Type) [Field K]
+    [Algebra k K] :
+    Function.Injective (Algebra.TensorProduct.map (AlgHom.id K K) f) := by
+  apply Module.Flat.lTensor_preserves_injective_linearMap
+  exact h
+
+
+/-- The k-algebra A is geometrically reduced iff its basechange to AlgebraicClosure k is reduced -/
+@[mk_iff]
+class IsGeometricallyReduced : Prop where
+  baseChangeReduced : IsReduced (TensorProduct k (AlgebraicClosure k) A)
+
+theorem IsGeometricallyReduced_imp_baseChange_by_closure_Reduced (h : IsGeometricallyReduced k A) :
+    IsReduced (TensorProduct k (AlgebraicClosure k) A) := by
+    exact h.baseChangeReduced
+
+lemma isGeometricallyReduced_of_injective (B : Type) [CommRing B] [Algebra k B] (f : A →ₐ[k] B)
+    (hf : Function.Injective f) [IsGeometricallyReduced k B] : IsGeometricallyReduced k A := by
+  let fK := baseChange  k A B f (AlgebraicClosure k)
+  have hfK : Function.Injective fK := baseChange_by_field_inj k A _ f hf _
+  expose_names
+  rw [isGeometricallyReduced_iff] at *
+  exact isReduced_of_injective fK hfK
+
+
+theorem isGeometricallyReduced_isReduced [IsGeometricallyReduced k A] : IsReduced A := by
```

Comment:
```suggestion
theorem isReduced_of_isGeometricallyReduced [IsGeometricallyReduced k A] : IsReduced A := by
```
Remember that the main statement always goes first, followed by the assumptions separated by `of`.